src/privatesend-util.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2014-2017 The Dash Core developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#include "privatesend-util.h"
+
+KeyHolder::KeyHolder(CWallet* pwallet)
+        :reserveKey(pwallet)
+{
+        reserveKey.GetReservedKey(pubKey, false);
+}
+
+void KeyHolder::KeepKey()
+{
+    reserveKey.KeepKey();
+}
+
+void KeyHolder::ReturnKey()
+{
+    reserveKey.ReturnKey();
+}
+
+CScript KeyHolder::GetScriptForDestination()
+{
+    return ::GetScriptForDestination(pubKey.GetID());
+}
+
+
+KeyHolderPtr KeyHolderStorage::AddKey(CWallet* pwallet)
```

Comment:
`AddKey` is only used in one place so far, and the return value is only used to extract the script.

`CScript scriptDenom = keyHolderStorage.AddKey(pwalletMain)->GetScriptForDestination();`

So, perhaps return the `CScript` directly rather than a pointer or handle to a KeyHolder.

I see that the shared pointer would be necessary if multiple asynchronous callers were storing handles.
That is not the case with the current code so I'd suggest to simplify the storage until this is needed.